IPython/core/completer.py

```diff
@@ -1879,6 +1891,14 @@ def _complete(self, *, cursor_line, cursor_pos, line_buffer=None, text=None,
         self.line_buffer = line_buffer
         self.text_until_cursor = self.line_buffer[:cursor_pos]
 
+        # Do magic arg matches
+        for matcher in self.magic_arg_matchers:
+            matches = [(m, matcher.__qualname__) for m in matcher(line_buffer)]
```

Comment:
Right, but you never mix entries from the two matchers. You could achieve the same thing like this:

```python
for matcher in self.magic_arg_matchers:
    matches = matcher(line_buffer)
    if matches:
        origins = [matcher.__qualname__] * len(matches)
        return text, matches, origins, ()
```

i.e. skip putting the origins into a list-of-tuples and then separating them back out.